### PR TITLE
own: match linked code host handles.

### PR DESCRIPTION
**Pull request by sashaostrikov**
Test plan:
Unit tests added.

Closes https://github.com/sourcegraph/sourcegraph/issues/52142

### CODE REVIEW

File: enterprise/internal/own/ownref_test.go [hunk 1]
```diff
@@ -81,6 +89,20 @@ func TestSearchFilteringExample(t *testing.T) {
...
+			RepoContext: &RepoContext{
+				Name:         "github.com/sourcegraph/sourcegraph",
+				CodeHostKind: "github",
+			},
+		},
+		"@jdoegl (gitlab handle) entry in CODEOWNERS": {
+			Handle: gitLabLogin,
+			RepoContext: &RepoContext{
+				Name:         "github.com/sourcegraph/sourcegraph",
+				CodeHostKind: "github",
```
Comment (cbart) [inline enterprise/internal/own/ownref_test.go:103]:
```suggestion
				CodeHostKind: "gitlab",
```

File: enterprise/internal/own/ownref_test.go [hunk 2]
```diff
@@ -290,3 +305,50 @@ func TestBagUnverifiedEmailOnlyMatchesWithItself(t *testing.T) {
 		})
 	}
 }
+
+func initUser(ctx context.Context, t *testing.T, db edb.EnterpriseDB) (*types.User, error) {
```
Comment (cbart) [inline enterprise/internal/own/ownref_test.go:309]: `t.Helper()` will help more precise error reporting

```suggestion
func initUser(ctx context.Context, t *testing.T, db edb.EnterpriseDB) (*types.User, error) {
  t.Helper()
```

**Review by cbart — APPROVED** [review]
🚢

File: enterprise/internal/own/ownref.go [hunk 3]
```diff
@@ -169,6 +161,46 @@ func ByEmailReference(ctx context.Context, db edb.EnterpriseDB, email string) (B
...
+	}
+	codeHostHandles := make([]string, 0, len(accounts))
+	for _, account := range accounts {
+		p := providers.GetProviderbyServiceType(account.ServiceType)
+		if p == nil {
+			return nil, errors.Errorf("cannot find authorization provider for the external account, service type: %s", account.ServiceType)
+		}
+		data, err := p.ExternalAccountInfo(ctx, *account)
+		if err != nil || data == nil {
+			return nil, err
```
Comment (coury-clark) [inline enterprise/internal/own/ownref.go:194]:
```suggestion
			return nil, errors.Wrap(err, "ExternalAccountInfo")
```

**Review by coury-clark — APPROVED** [review]

File: enterprise/internal/own/ownref.go [hunk 4]
```diff
@@ -169,6 +161,46 @@ func ByEmailReference(ctx context.Context, db edb.EnterpriseDB, email string) (B
...
+		}
+		if data.Login != nil && len(*data.Login) > 0 {
+			codeHostHandles = append(codeHostHandles, *data.Login)
+		}
+	}
+	b = append(b, &userReferences{codeHostHandles: codeHostHandles})
```
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply enterprise/internal/own/ownref.go:200]: that's a great catch, I agree, incorporating the change 👍🏻

File: enterprise/internal/own/ownref.go [hunk 5]
```diff
@@ -177,8 +209,10 @@ type userReferences struct {
 	user *types.User
 	// handle text input used to search that reference,
 	// it is present if user was not found in the database.
-	handle         string
-	verifiedEmails []string
+	handle string
+	// codeHostHandles are user handles of connected code hosts of a resolved user.
+	codeHostHandles []string
```
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply enterprise/internal/own/ownref.go:214]: hmm, I cannot think of a use/corner case at the moment, maybe when we wire it up with frontend, it becomes more obvious

File: enterprise/internal/own/ownref_test.go (hunk 2 above)
Comment (sashaostrikov) [reply enterprise/internal/own/ownref_test.go:309]: yes, forgot to add, adding now

File: enterprise/internal/own/ownref_test.go [hunk 6]
```diff
@@ -208,6 +223,13 @@ func TestBagUserFoundNoMatches(t *testing.T) {
 				CodeHostKind: "github",
 			},
 		},
+		"different code host handle entry in CODEOWNERS": {
+			Handle: gerritLogin,
+			RepoContext: &RepoContext{
+				Name:         "github.com/sourcegraph/sourcegraph",
+				CodeHostKind: "github",
```
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply enterprise/internal/own/ownref_test.go:230]: yep, inaccurate test, fixed!

**Pull request merged**